Mathlib/Analysis/InnerProductSpace/Simultaneous.lean

```diff
@@ -0,0 +1,113 @@
+/-
+Copyright (c) 2024 Jon Bannon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jon Bannon, Jack Cheverton, Samyak Dhar Tuladhar
+-/
+
+import Mathlib.Analysis.InnerProductSpace.Spectrum
+import Mathlib.Analysis.InnerProductSpace.Projection
+import Mathlib.Order.CompleteLattice
+import Mathlib.LinearAlgebra.Eigenspace.Basic
+
+/-! # Simultaneous eigenspaces of a commuting pair of symmetric operators
+
+This file collects various decomposition results for simultaneous eigenspaces of a commuting pair
+of symmetric operators on a finite-dimensional inner product space.
+
+# Main Result
+
+* `DirectSum.IsInternal.eigenspaces_of_commuting_symmetric_pair` establishes that
+   if `{A B : E →ₗ[𝕜] E}`, then `IsSymmetric A`, `IsSymmetric B` and `A ∘ₗ B = B ∘ₗ A` imply that
+   `E` decomposes as an internal direct sum of the pairwise orthogonal spaces
+   `eigenspace B μ ⊓ eigenspace A ν`
+
+## TODO
+
+Develop a `Diagonalization` structure for linear maps and / or matrices which consists of a basis,
+and a proof obligation that the basis vectors are eigenvectors.
+
+## Tags
+
+self-adjoint operator, simultaneous eigenspaces, simultaneous diagonalization
+
+-/
+
+variable {𝕜 : Type*} [RCLike 𝕜]
+variable {E : Type*} [NormedAddCommGroup E] [InnerProductSpace 𝕜 E] [FiniteDimensional 𝕜 E]
+
+open Module.End
+
+namespace LinearMap
+
+namespace IsSymmetric
+
+section Pair
+
+variable {α : 𝕜} {A B : E →ₗ[𝕜] E} (hA : A.IsSymmetric) (hB : B.IsSymmetric)
+    (hAB : A ∘ₗ B = B ∘ₗ A)
+
+/--If a pair of operators commute, then the eigenspaces of one are invariant under the other.-/
+theorem eigenspace_invariant_of_commute (α : 𝕜) :
+    ∀ v ∈ (eigenspace A α), (B v ∈ eigenspace A α) := by
+  intro v hv
+  rw [eigenspace, mem_ker, sub_apply, Module.algebraMap_end_apply, ← comp_apply A B v, hAB,
+  comp_apply B A v, ← map_smul, ← map_sub, hv, map_zero] at *
+
+/--The inf of an eigenspace of an operator with another invariant subspace
+agrees with the corresponding eigenspace of the restriction of that operator to the
+invariant subspace-/
+theorem eigenspace_restrict_eigenspace_eq_inf : (fun (γ : 𝕜) ↦
+    Submodule.map (Submodule.subtype (eigenspace A α)) (eigenspace (B.restrict
+    (eigenspace_invariant_of_commute hAB α)) γ))
+    = (fun (γ : 𝕜) ↦ (eigenspace A α ⊓ eigenspace B γ)) := by
+  funext γ
+  exact Eq.symm (Submodule.inf_genEigenspace B (eigenspace A α)
+    (eigenspace_invariant_of_commute hAB α) (μ := γ) (k := 1))
```

Comment:
I don't think you want the function version. Just pass `γ` as a parameter, like below. Note that I have also reversed the direction of the equality, and we can simplify the name that way.
```
open Submodule in
/-- The intersection of eigenspaces of commuting selfadjoint operators is equal to the eigenspace of
one operator restricted to the eigenspace of the other, which is an invariant subspace because the
operators commute. -/
theorem eigenspace_inf_eigenspace (α γ : 𝕜) :
    eigenspace A α ⊓ eigenspace B γ = map (Submodule.subtype (eigenspace A α))
      (eigenspace (B.restrict (eigenspace_invariant_of_commute hAB α)) γ) :=
  (eigenspace A α).inf_genEigenspace _ _ (k := 1)
```